onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorConcat.cpp

```diff
@@ -78,6 +78,25 @@ class DmlOperatorConcat : public DmlOperator, public ConcatHelper
     }
 };
 
+bool CALLBACK QueryGraphSupportConcat(
+    const Windows::AI::MachineLearning::Adapter::EdgeShapes& inputShapes,
+    const Windows::AI::MachineLearning::Adapter::EdgeShapes& outputShapes)
+{
+    // Concat is always supported in graphs as long as at least one input is not empty. Empty inputs are ignored and
+    // the rest are concatenated together.
+    for (uint32_t i = 0; i < inputShapes.EdgeCount(); ++i)
+    {
+        const std::vector<uint32_t>& shape = inputShapes.GetShape(i);
+
```

Comment:
[minor] Consider adding `begin`/`end` methods to `EdgeShapes` so you can enjoy ranged `for`.

```
    for (auto& shape : inputShapes)
    {
```